### ISSUE TITLE
TTML Cues not translated to VTT correctly

### ISSUE THREAD

**Issue by sandeep-mirpuri** [google/shaka-player#1020]
**Shaka Version 2.2.1, Chrome Version 60 on MacOS Sierra**

**What did you do?**

I am trying to display TTML closed captions on a DASH asset with closed captions within the stream. The TTML specifies the origin, extent and displayAlign properties for cues.

**What did you expect to happen?**

The expected result was that Shaka's ttml_text_parser would convert the cues into WebVTT cues and display it correctly at the bottom center.

**What actually happened?**

The cues appear in the top left corner of the screen. I am aware that displayAlign is not a VTT cue property, but there are other properties which don't seem to be getting parsed correctly. By following the code I noted that these are being added in a function belonging to **shaka.text.TtmlTextParser** called **addStyle_()**. In this function the line, size, snapToLines, position and positionAlign properties get set depending on whether the origin and extent are specified as a percent - 

```
results = TtmlTextParser.percentValues_.exec(extent);
if (results != null) {
    cue.size = Number(results[1]);
}
```

and 

```
results = TtmlTextParser.percentValues_.exec(origin);
if (results != null) {
    if (cue.writingDirection ==
         Cue.writingDirection.HORIZONTAL_LEFT_TO_RIGHT ||
         cue.writingDirection ==
         Cue.writingDirection.HORIZONTAL_RIGHT_TO_LEFT) {
        cue.position = Number(results[1]);
        cue.line = Number(results[2]);
      } else {
        cue.position = Number(results[2]);
        cue.line = Number(results[1]);
      }
    cue.lineInterpretation = Cue.lineInterpretation.PERCENTAGE;
}
```

However, when I tried overriding these to 
line = "auto";
position = "auto";
positionAlign = "auto";
size = 100;
snapToLines = true;

Then the captions display correctly.

So either the properties are not being translated to VTT correctly or VTT can't deal with these properties and it works better if we leave it to the browser. Can this be handled in any way?

Please see these screenshots - 

**Before the overrides**
<img width="1440" alt="before properties override" src="https://user-images.githubusercontent.com/30884334/30359471-45886580-988e-11e7-85b7-3d081fe045b9.png">

**After the overrides**
<img width="1440" alt="after properties override" src="https://user-images.githubusercontent.com/30884334/30359470-4584a1ca-988e-11e7-857e-9888e97a1a79.png">

**Comment by TheModMaker** [comment]
Do you have some example content that we could use?  It looks like your overrides are just removing the attributes in the TTML.  If the TTML says the origin should be in the top-left, then that's where they should appear.  Do you want us to ignore the style in the TTML or are we parsing the attributes incorrectly?

**Comment by sandeep-mirpuri** [comment]
@TheModMaker yes you are right. My overrides are intentionally removing the attributes because it places the cues better. The reason I feel that it may not be being parsed correctly is because of what the resulting VTT cue properties look like. Unfortunately I can't share the content. It is protected. But I can tell you what specs are coming in from the TTML.

So in the TTML we get - 

`tts:origin = "10% 10%", tts:extent = "80% 80%", displayAlign = "after"`

Which basically means that there should be a text track container 80% height and width, centered in the video container and the text should be displayed at the bottom of this container. However, based on the TTML parser code this translates to - 

`cue.size = 80, cue.position = 10, cue.line = 10, cue.snapToLines = false`

Which explains why it looks like it does in the screen shot. I am not sure if this is correct and I feel like the displayAlign makes all the difference. So while VTT cues do not support all the properties of TTML, I was wondering if there is a better solution than removing the TTML attributes and forcing it to "auto"?

**Comment by TheModMaker** [comment]
There is a better solution, but no browser implements it.  There exists a `VTTRegion` that is used just for this.  It allows specifying a sub-region to render the cues into.  However, as of now, there are no browsers that implement it.

So I think we will parse those attributes and expose them in our new `shaka.text.Cue` interface so a custom text renderer could handle it.  In the default renderer, we will try to add the `VTTRegion`; but for now, it will just ignore them.

**Comment by sandeep-mirpuri** [comment]
@TheModMaker is it possible to add an option to default to the auto if the browser doesn't support VTTRegion and the developer does not necessarily want to implement a custom text renderer?

**Comment by TheModMaker** [comment]
That will be the default.  We will effectively ignore the setting and use the browser defaults (which is 'auto').

**Comment by joeyparrish** [comment]
The fix has been cherry-picked to v2.2.x and will be released in v2.2.2.

**Comment by sandeep-mirpuri** [comment]
Thanks guys
